README.md
# **Trail**
[![NPM version](https://img.shields.io/npm/v/@dinoly/trail/latest?color=blue&label=&logo=npm)](https://www.npmjs.com/package/@dinoly/trail)
[![Github action](https://img.shields.io/github/workflow/status/dinoly/trail/Quality?label=CI&logo=github)](https://github.com/dinoly/trail/actions/workflows/main.yml)
[![License](https://img.shields.io/github/license/dinoly/trail?color=white)](https://github.com/dinoly/trail/blob/main/LICENSE)

[![Milestones](https://img.shields.io/github/milestones/progress/dinoly/trail/1?style=social)](https://github.com/dinoly/trail/milestone/1)

**NOTE: There are some breaking changes in 0.4.0**

Trail, Generates a trail behind HTMLElement and in HTMLCanvasElement.

look into the [Change log](./CHANGELOG.md) for features and breaking changes

## Content Outline
- [Introduction](#introduction)
- [HTML Trail](#html-trail)
  - [Sample code](#sample-code)
  - [Arguments](#arguments)
  - [Methods](#methods)
- [Canvas Trail](#canvas-trail)
  - [Canvas Sample code](#canvas-sample-code)
  - [Canvas Arguments](#canvas-arguments)
  - [Canvas Methods](#canvas-methods)
- [Gifs](#gifs)
- [Bugs and Future Improvements](#bugs-and-future-improvements)

## Introduction
Place the following `<script>` near the end of your pages, right before the closing `</body>` tag.

```html
<script src="https://cdn.jsdelivr.net/npm/@dinoly/trail@0.4.0/dist/trail.min.js" crossorigin="anonymous"></script>
```
After the above script, make a new trail object and pass in the **class** of the HTMLElement as a `target` parameter.
```html
<script>
  const move = new Trail({
    target: "circle"
  });
  move.followMouse();
</script>
```
Or create trails using Canvas, pass in the **id** of the HTMLCanvasElement as a `area` parameter.
```html
<script>
const can = new CanvasTrail({
  area: "canvas",
  color: "#ffffff50",
  effect: "line"
})
can.followMouse();
</script>
```
## HTML Trail
### Sample code
```html
<!-- index.html -->
<!DOCTYPE html>
<html lang="en" dir="ltr">
<head>
  <meta charset="utf-8">
  <title>Trail</title>
  <link rel="stylesheet" href="./styles.css" /> <!-- Link to the style sheet with circle's custom styling  -->
</head>

<body>
  <div class="circle"></div>
  <script src="https://cdn.jsdelivr.net/npm/@dinoly/trail@0.4.0/dist/trail.min.js" crossorigin="anonymous"></script>
  <script>
    const move = new Trail({
      target: "circle",
    });
    move.followMouse();
  </script>
</body>

</html>
```
```css
/* styles.css */
body {
  margin: 0;
  padding: 0;
  background: black;
  width: 100vw;
  height: 100vh;
}
* {
  cursor: none;
}
.circle {
  background-color: lightblue;
  width: 1.5rem;
  height: 1.5rem;
  border-radius: 50%;
  pointer-events: none;
  margin: 10px; /* will be removed */
}
```

### Arguments
```js
{
  target: null,
  area: null,
  particle: "default", // self
  offset: ['default', 'default'], // 0px
  delay: 50, // in milliseconds
  color: "default", // white
  isnode: true,
  effect: "default", // straight
  trails: false
}
```

| Name         | Description     | code | types |
|--------------|-----------|------------|-------|
| **target** | Is the **class** of the HTMLElement behind which the trail will be generated. | `target: "example"`| none |
| **area** | Is the **class** of the element inside which the trail will be generated. | `area: "example"`| none |
| **particle** | Shape of the particle in trail, default: `self`. Additionally one can provide the particle style instead of **pre-defined types**| `particle: "circle"` or `particle: "height:.3rem;width:.3rem; background-color:salmon;",`| "circle", "triangle", "square" |
| **offset** | Offsets the trail from origin | `color: "mediumseagreen"`| none |
| **color** | Color of the trail. Can be a gradient | `color: "#ffffff50"` or `color: [[0,"orange"], [30,"white"], [50, "green"]]`| none |
| **delay** | Delay particle emission | `delay: 1000 // 1s`| none |
| **isnode** | Set `false` if you want a trail for a **Text** or **Svg**, default: `true` | `isnode:false`| none |
| **effect** | Behavior of the particle | `effect:"spread"`| "spread", "rotate" |
| **trails** | If `true` will create trails with multiple particles, currently `2`. | `trails:true`| none |


Note: if this `trails:true` creates a lag, use the `effect: "spread"` which will create similar effect with one particle.


### Methods
| Name         | Description     | code
|--------------|   -----------   |------------|
| **followMouse** | Make the Html element follow the **Mouse Pointer** with a trail. | `Trail.followMouse();` |
| **followNode** | Does not make the Html element follow the **Mouse Pointer**, add CSS to move the element how ever you like. | `Trail.followNode();` |
| **activeArea** | make the Html element follow the **Mouse Pointer** when mouse hover a certain element, set with `area: <class-of-area>` | `Trail.activeArea();` |

## Canvas Trail
### Canvas sample code
```html
<!-- index.html -->
<!DOCTYPE html>
<html lang="en" dir="ltr">
<head>
  <meta charset="utf-8">
  <title>Trail</title>
  <link rel="stylesheet" href="./styles.css" /> <!-- Link to the style sheet with circle's custom styling  -->
</head>

<body>
  <canvas id="canvas"></canvas>
  <script src="https://cdn.jsdelivr.net/npm/@dinoly/trail@0.4.0/dist/trail.min.js" crossorigin="anonymous"></script>
  <script>
  const can = new CanvasTrail({
    area: "canvas",
    color: "#ffffff50",
    effect: "paint"
  })
  can.followMouse();
  </script>
</body>

</html>
```
```css
/* styles.css */
body {
  margin: 0;
  padding: 0;
  background: black;
  width: 100vw;
  height: 100vh;
}
* {
  cursor: none;
}
#canvas {
  width: 100%;
  height: 100%;
  background: linear-gradient(#25364f, #4d71a5, #9bc4ff);
}
```


### Canvas Arguments
```js
{
  area: null,
  color: "black",
  effect: "default",
  size: 5
}
```

| Name         | Description     | code | types |
|--------------|-----------|------------|-------|
| **area** | Is the **id** of the HTMLCanvasElement inside which the trail will be generated. | `area: "canvas"`| none |
| **color** | Color of the trail | `color:"#ffffff50"`| none |
| **particle** | Behavior of the trail | `particle:"paint"`| "default", "paint", "web", "line" |
| **size** | size of the particle, ranges between `1 to 9` | `size:"2"`| none |


### Canvas Methods
| Name         | Description     | code
|--------------|   -----------   |------------|
| **followMouse** | Make the HTMLCanvasElement follow the **Mouse Pointer** with a trail. | `CanvasTrail.followMouse();` |

### Gifs
<div style="display:flex;flex-direction:row;width:400px;">
  <img src="./followMouse.gif" alt="follow mouse" width="250px"/>
  <img src="./followNode.gif" alt="follow mouse" width="250px"/>
  <img src="./activeArea.gif" alt="follow mouse" width="250"/>
  <img src="./gradient.gif" alt="Gradient Trail" width="250"/>
</div>

### Bugs and Future Improvements
Improvements
- Resizing the trail.
- Trail behind multiple elements with one `Trail` object.
- More Effects and Particles.
- Support for nested elements.


CHANGELOG.md
# Changelog
## 0.4.0 (2021-11-29)
#### Merged pull requests:
- Add custom styling to particle [#11](https://github.com/dinoly/trail/pull/11)
- Added delay and offset parameter [#12](https://github.com/dinoly/trail/pull/12)
- Gradient color in Trail [#14](https://github.com/dinoly/trail/pull/14)
- **CanvasTrail** Class [#19](https://github.com/dinoly/trail/pull/19)
- Changelog [#21](https://github.com/dinoly/trail/pull/21)

#### Bug fixed:
- Fixed particle spread effect [#17](https://github.com/dinoly/trail/pull/17)
- Fixed Text offset and triangle particle color [#18](https://github.com/dinoly/trail/pull/18)

## 0.3.6 (2021-09-04)
#### Merged pull requests:
- corrected version in readme [#8](https://github.com/dinoly/trail/pull/8)

## 0.3.3 (2021-09-04)
#### Merged pull requests:
- Fixed styling [#7](https://github.com/dinoly/trail/pull/7)

## 0.3.2 (2021-09-01)
#### Merged pull requests:
- Fixed trail offset [#6](https://github.com/dinoly/trail/pull/6)

## 0.3.1 (2021-08-26)
Changed javaScript with typeScript.
#### Merged pull requests:
- Workflow and `target` element offset [#5](https://github.com/dinoly/trail/pull/5)

## 0.3.0 (2021-08-23)
#### Merged pull requests:
- New method **activeArea** [#4](https://github.com/dinoly/trail/pull/4)

## 0.2.0 (2021-08-20)
#### Changes
- Changed `class` argument to `target`

#### Implemented enhancements:
- New argument: `color`
- New effect: `rotate` [bbbb0a0](https://github.com/dinoly/trail/pull/3/commits/bbbb0a09c17f560440f127165fb0330e77471638)

#### Merged pull requests:
- New effect and some bug fixes [#3](https://github.com/dinoly/trail/pull/3)
- Transparent the background of **triangle** particle [#2](https://github.com/dinoly/trail/pull/2)

## 0.1.0 (2021-08-18)
#### Merged pull requests:
- Initial logic for the trail [#1](https://github.com/dinoly/trail/pull/1)


dev/trail.js
var circle = '_trail{background-color:white;width:0.3rem;height:0.3rem;border-radius:50%}';
var triangle ='_trail{width:0;height:0;background-color:transparent;border-left:0.25rem solid transparent;border-right:0.25rem solid transparent;border-bottom:0.25rem solid white}';
var square = '_trail{background-color:white;width:0.3rem;height:0.3rem;border-radius:0}';
var singleT = '.anim{animation:disappear 1s ease-out forwards}@keyframes disappear{0%{opacity:1}100%{opacity:0}}';

class Trail {
  #target = "";#particle = "";#color = "";#effect = "";#isnode = "";#trails="";#styles="";#area="";#bounds="";
  constructor(props){
    this.#target = props.target;
    this.node = document.querySelector('.'+this.#target);
    this.#bounds = this.node.getBoundingClientRect();
    this.#isnode = props.isnode === false ? props.isnode : true;
    this.area = props.area;
    this.#color = props.color;
    this.margin = props.margin ? props.margin : '0px';
    this.#particle = props.particle ? props.particle : 'self';
    this.tick = props.tick ? props.tick * 1000 : 5000;
    this.#styles = props.styles ? props.styles : singleT;
    this.#effect = props.effect ? props.effect : "straight";
    this.#trails = props.trails === true ? props.trails : false;
    this.#setUpStyles();
    this.#setUpParticles();
  }

  #setUpStyles(){
    this.node.style.position = 'absolute';
    this.node.style.transform = `translate(calc(-100% - ${this.margin}), calc(-100% - ${this.margin}))`;
    this.node.style.zIndex = `10000`;
    this.node.style.pointerEvents = "none";
    this.node.style.margin = "0";
  }

  #setUpParticles(){
    if(this.#particle === 'circle'){
      this.#styles += "." + this.#target + circle;
      if(this.#color){
        this.#styles += "." + this.#target + `_trail{background: ${this.#color};color:${this.#color};`
      }
    }
    else if(this.#particle === 'triangle'){
      this.#styles += "." + this.#target + triangle;
      if(this.#color){
        this.#styles += "." + this.#target + `_trail{border-bottom:0.25rem solid ${this.#color};color:${this.#color};`
      }
    }
    else if(this.#particle === 'square'){
      this.#styles += "." + this.#target + square;
      if(this.#color){
        this.#styles += "." + this.#target + `_trail{background: ${this.#color};color:${this.#color};`
      }
    }
    let stylesheet = document.createElement('style');
    stylesheet.type = 'text/css';
    stylesheet.innerHTML = this.#styles;
    document.head.appendChild(stylesheet);
  }

  followMouse(){
    console.log(this.node.getBoundingClientRect().width, this.node.getBoundingClientRect().height)
    document.addEventListener("mousemove",(pos) =>{
      this.node.style.left = (pos.clientX + this.#bounds.width/2) + 'px';
      this.node.style.top = (pos.clientY + this.#bounds.height/2) + 'px';
      this.#trails ? this.#createParticles() : this.#createParticle();
    })
  }

  followNode(){
    var that = this;
    setInterval(function(){
      let pos = that.node.getBoundingClientRect();
      that.node.style.left = (pos.left + that.#bounds.width) + 'px';
      that.node.style.top = (pos.top + that.#bounds.height) + 'px';
      that.#trails ? that.#createParticles() : that.#createParticle();
    }, 100)
  }

  activeArea(){
    let _area = document.querySelector('.'+this.area);
    _area.addEventListener('mouseover', ()=>{
      _area.addEventListener("mousemove",(pos) =>{
        this.node.style.left = (pos.clientX + this.#bounds.width/2) + 'px';
        this.node.style.top = (pos.clientY + this.#bounds.height/2) + 'px';
        this.#trails ? this.#createParticles() : this.#createParticle();
      })
    })

    _area.addEventListener('mouseout', ()=>{
      _area.removeEventListener("mousemove",()=>{})
    })
  }

  #createParticle(rand=0){
    let randV
    if(this.#effect === "spread"){
      randV = Math.floor(Math.random()*12-6);
    }else{
      randV = rand;
    }
    let fy;
    if(this.#isnode === false){
      fy = this.node.cloneNode(true);
    }
    else{
      fy = this.node.cloneNode(false);
    }
    if(this.#effect === "rotate"){
      let randA = Math.floor(Math.random()*120+30).toString()+"deg";
      fy.style.transform += `rotate(${randA})`;
    }
    if(this.#particle !== "self"){
      fy.classList.remove(this.target);
    }
    fy.classList.add("anim", `${this.#target}_trail`);
    fy.style.left = (Number(this.node.style.left.replace("px", ""))+randV)+"px";
    fy.style.top = (Number(this.node.style.top.replace("px", ""))+randV)+"px";
    if(this.#particle !== "self"){
      fy.style.left = (Number(this.node.style.left.replace("px", ""))+randV-this.#bounds.width/3)+"px";
      fy.style.top = (Number(this.node.style.top.replace("px", ""))+randV-this.#bounds.height/3)+"px";
    }
    fy.style.zIndex = "0";
    document.body.appendChild(fy);

    setTimeout(function () {
      fy.addEventListener("animationend", fy.parentNode.removeChild(fy));
    }, 1000);
  }

  #createParticles(){
    for(let i=0;i<2;i++){
      let randV = Math.floor(Math.random()*12-6);
      this.#createParticle(randV);
    }
  }
}


src/trail.ts
/*!
  Dinoly v0.3.1 (https://www.npmjs.com/package/@dinoly/trail)
  Licensed under MIT (https://github.com/dinoly/trail/blob/main/LICENSE)
*/

const circle = "_trail{background:white;width:0.3rem;height:0.3rem;border-radius:50%}";
const triangle ="_trail{width:0;height:0;background:transparent;border-left:0.25rem solid transparent;border-right:0.25rem solid transparent;border-bottom:0.25rem solid white}";
const square = '_trail{background:white;width:0.3rem;height:0.3rem;border-radius:0}';
const singleT = '.anim{animation:disappear 1s ease-out forwards, grad 1s ease-out forwards}@keyframes disappear{0%{opacity:1}100%{opacity:0}}';

interface itrail{
  target: string;
  particle?: string;
  color?: string | Map<number, string>;
  effect?: string;
  isnode?: boolean;
  trails?: boolean;
  styles: string;
  area?: string;
  bounds:DOMRect;
  node:HTMLElement;
  offset?: string[];
  delay?: number;
}


// eslint-disable-next-line @typescript-eslint/no-unused-vars
class Trail {
  #target:string;node:HTMLElement;#particle:string;#color:string|Map<number, string>;#effect:string;#isnode:boolean;#trails:boolean;#styles:string;#area?:string;#bounds:DOMRect;#offset:string[];#delay:number;
  constructor(props:itrail){
    this.#target = props.target;
    this.node = document.querySelector('.'+this.#target)!;
    this.#bounds = this.node.getBoundingClientRect();
    this.#isnode = props.isnode === false ? props.isnode : true;
    this.#area = props.area;
    this.#color = props.color ? typeof(props.color) === "function"? new Map(props.color) : props.color: "";
    this.#offset = props.offset ? props.offset : ['0px', '0px'];
    this.#particle = props.particle ? props.particle : "self";
    this.#delay = props.delay ? props.delay : 50;
    this.#styles = props.styles ? props.styles : singleT;
    this.#effect = props.effect ? props.effect : "straight";
    this.#trails = props.trails === true ? props.trails : false;
    this.#setUpStyles();
    this.#setUpParticles();
  }

  #setUpStyles():void{
    this.node.style.position = 'fixed';
    this.node.style.transform = `translate(calc(-100% + ${this.#offset[0]}), calc(-100% - ${this.#offset[1]}))`;
    this.node.style.zIndex = `10000`;
    this.node.style.pointerEvents = "none";
    this.node.style.margin = "0";
  }

  #setUpParticles():void{
    if(this.#particle === 'circle'){
      this.#styles += "." + this.#target + circle;
    }
    else if(this.#particle === 'triangle'){
      this.#styles += "." + this.#target + triangle;
    }
    else if(this.#particle === 'square'){
      this.#styles += "." + this.#target + square;
    }
    else{
      this.#styles += "." + this.#target + `_trail{${this.#particle}}`;
    }
    if(typeof(this.#color) === "string"){
      if(this.#color.length > 0){
        this.#styles += "." + this.#target + (this.#particle !== "triangle" ? `_trail{background: ${this.#color};color:${this.#color};` : `_trail{background:transparent;border-bottom:0.25rem solid ${this.#color};color:${this.#color};`);
      }
    }
    else if(typeof(this.#color) === "object"){
      let grad = ""
      this.#color.forEach((value):void => {
        grad += `${value[1]}%{background-color:${value[0]}}`
      });
      this.#styles += `@keyframes grad{${grad}}`
    }
    const stylesheet = document.createElement('style');
    // stylesheet.type = 'text/css';
    stylesheet.innerHTML = this.#styles;
    document.head.appendChild(stylesheet);
  }

  followMouse():void{
    document.addEventListener("mousemove",(pos) =>{
      this.node.style.left = (pos.clientX + this.#bounds.width/2) + 'px';
      this.node.style.top = (pos.clientY + this.#bounds.height/2) + 'px';
    })
    setInterval(() => {
      this.#trails ? this.#createParticles() : this.#createParticle();
    }, this.#delay)
  }

  followNode():void{
    setInterval(() => {
      const pos = this.node.getBoundingClientRect();
      this.node.style.left = (pos.left + this.#bounds.width) + 'px';
      this.node.style.top = (pos.top + this.#bounds.height) + 'px';
      this.#trails ? this.#createParticles() : this.#createParticle();
    }, this.#delay)
  }

  activeArea():void{
    const _area = document.querySelector('.'+this.#area)!;
      _area.addEventListener('mouseover', ()=>{
          _area.addEventListener("mousemove",((pos:MouseEvent) =>{
            this.node.style.left = (pos.clientX + this.#bounds.width/2) + 'px';
            this.node.style.top = (pos.clientY + this.#bounds.height/2) + 'px';
            setTimeout(() => {
            this.#trails ?  this.#createParticles() : this.#createParticle();
            }, 5000);
          }) as EventListener);
      })

      _area.addEventListener('mouseout', ()=>{
          _area.removeEventListener("mousemove",()=>{});
      })
  }

  #createParticle(rand={"x": 0,"y": 0}):void{
      const randV:{"x": number, "y": number} = rand;
      if(this.#effect === "spread"){
        randV.x = Math.floor(Math.random()*12-6);
        randV.y = Math.floor(Math.random()*12-6);
      }
      let newP:any;
      if(this.#isnode === false){
        newP = this.node.cloneNode(true);
      }
      else{
        newP = this.node.cloneNode(false);
      }
      if(this.#effect === "rotate"){
        const randA = Math.floor(Math.random()*120+30).toString()+"deg";
        newP.style.transform += `rotate(${randA})`;
      }
      if(this.#particle !== "self"){
        newP.classList.remove(this.#target);
      }
      newP.classList.add("anim", `${this.#target}_trail`);
      if(this.#particle !== "self"){
        newP.style.left = (Number(this.node.style.left.replace("px", ""))+randV.x - this.#bounds.width/2 + 2.4)+"px";
        newP.style.top = (Number(this.node.style.top.replace("px", ""))+randV.y - this.#bounds.height/2 + 2.4)+"px";
      }
      else{
        newP.style.left = (Number(this.node.style.left.replace("px", ""))+randV.x)+"px";
        newP.style.top = (Number(this.node.style.top.replace("px", ""))+randV.y)+"px";
      }
      newP.style.zIndex = "0";
      document.body.appendChild(newP);

      setTimeout(function () {
        newP.addEventListener("animationend", newP.parentNode.removeChild(newP));
      }, 1000);
  }

  #createParticles():void{
    for(let i=0;i<2;i++){
      const randV = {"x": Math.floor(Math.random()*12-6),"y": Math.floor(Math.random()*12-6)};
      this.#createParticle(randV);
    }
  }
}


// canvas particles
interface particle{
  x:number,
  y:number,
  size:number,
  range:number,
  color:string,
  weight:number,
  ctx:CanvasRenderingContext2D
  mouse:mouse
}
// types: Paint, Web, Line, Default
class Particle{
  x:number;y:number;size:number;range:number;color:string;weight:number;ctx:CanvasRenderingContext2D;mouse:mouse;
  #srink=0.1;
  constructor(props:particle){
    this.x = props.x;
    this.y = props.y;
    this.size = props.size;
    this.range = props.range;
    this.color = props.color;
    this.weight = props.weight;
    this.ctx = props.ctx;
    this.mouse = props.mouse;
  }
  draw():void{
    this.ctx.beginPath();
    this.ctx.arc(this.x, this.y, this.size, 0, Math.PI * 2, false);
    this.ctx.fillStyle = this.color;
    this.ctx.fill();
  }

  update(effect="default"):void{
    this.size -= this.#srink;
    if (this.size < 0){
      if (effect === "default"){
        this.x = this.mouse.x!;
        this.y = this.mouse.y!;
        this.size = (Math.random() * this.range) + this.range;
        this.weight = 1;
      }
      else if(effect==="paint"){
        this.x = (this.mouse.x! + ((Math.random() * 20) - 10));
        this.y = (this.mouse.y! + ((Math.random() * 20) - 10));
        this.size = (Math.random() * this.range) + this.range;
        this.weight = (Math.random() * 2) - 0.5;
      }
      else if(effect==="web"){
        this.x = (this.mouse.x! + ((Math.random() * 20) - 10));
        this.y = (this.mouse.y! + ((Math.random() * 20) - 10));
        this.size = (Math.random() * 3) + 2;
        this.weight = 1;
      }
      else if (effect==="line"){
        this.x = this.mouse.x!;
        this.y = this.mouse.y!;
        this.size = (Math.random() * 2) + 1;
        this.weight = 1;
      }
    }
  }
}


interface mouse{
  x:number | undefined;
  y:number | undefined;
}

interface ctrail{
  area: string,
  particle?: string,
  color?: string,
  effect?: string,
  size?: number
}


class CanvasTrail{
  #ctx:CanvasRenderingContext2D;canvas:HTMLCanvasElement;noOfParticles=80;particleArray:Particle[]=[];
  particle:string;color:string;effect:string;area:string;size:number;//offset:string[];delay:number;

  constructor(props:ctrail){
    this.area = props.area;
    this.canvas = <HTMLCanvasElement> document.getElementById(this.area)!;
    this.#ctx = this.canvas.getContext("2d")!;
    this.color = props.color ? props.color : "black";
    this.effect = props.effect ? props.effect : "default";
    this.particle = props.particle ? props.particle : 'self';
    // this.delay = props.delay ? props.delay : 50;
    this.particle = props.particle ? props.particle : "default";
    this.size = props.size ? props.size>0 && props.size < 10? props.size: 5 : 5;
    this.#setUpStyles();
  }

  #setUpStyles():void{
    this.canvas.style.position = 'absolute';
    // this.canvas.style.transform = `translate(calc(-100% + ${this.offset[0]}), calc(-100% - ${this.offset[1]}))`;
    this.canvas.style.zIndex = `-100`;
    this.canvas.style.pointerEvents = "none";
    this.canvas.width = window.innerWidth;
    this.canvas.height = window.innerHeight;
  }

  #setUpParticles(mouse:mouse):particle{
    const _particle = {
      x: Math.random() * this.canvas.width,
      y: Math.random() * this.canvas.width,
      size: 0.2, // (Math.random() * 5) + 2,
      color: this.color,
      range: this.size,
      weight: 0,
      mouse:mouse,
      ctx:this.#ctx
    }
    return _particle;
  }

  followMouse():void{
    const mouse:mouse={x:0, y:0};
    window.addEventListener("mousemove",(event)=>{
      mouse.x = event.x;
      mouse.y = event.y;
    });
    setInterval(function(){
      mouse.x = undefined;
      mouse.y = undefined;
    }, 200);

    this.animate();
    this.generateParticles(this.#setUpParticles(mouse));
  }
  generateParticles(particle:particle):void{
    this.particleArray = [];
    for(let i=0; i<this.noOfParticles; i++){
      this.particleArray.push(new Particle(particle));
    }
  }
  animate():void{
    this.#ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    for(let i=0; i<this.particleArray.length; i++){
      this.particleArray[i].update(this.particle);
      this.particleArray[i].draw();
    }
    if(this.particle==="web"){
      this.#connectWeb();
    }
    else if(this.particle==="line"){
      this.#connectLine();
    }
    requestAnimationFrame(this.animate.bind(this));
  }
  #connectWeb():void{
    // let opacityValue=1;
    for(let i=0; i<this.particleArray.length; i++){
      for(let j=i; j<this.particleArray.length; j++){
        const distance = ((this.particleArray[i].x - this.particleArray[j].x) * (this.particleArray[i].x - this.particleArray[j].x)) + ((this.particleArray[i].y - this.particleArray[j].y) * (this.particleArray[i].y - this.particleArray[j].y));
        if(distance < 400){
          // opacityValue = 1 - (distance/10000);
          this.#ctx.strokeStyle = this.color;
          this.#ctx.beginPath();
          this.#ctx.lineWidth = 1;
          this.#ctx.moveTo(this.particleArray[i].x, this.particleArray[i].y);
          this.#ctx.lineTo(this.particleArray[j].x, this.particleArray[j].y);
          this.#ctx.stroke();
        }
      }
    }
  }
  #connectLine():void{
    // let opacityValue=1;
    for(let i=0; i<this.particleArray.length-1; i++){
      // const distance = ((this.particleArray[i].x - this.particleArray[i+1].x) * (this.particleArray[i].x - this.particleArray[i+1].x)) + ((this.particleArray[i].y - this.particleArray[i+1].y) * (this.particleArray[i].y - this.particleArray[i+1].y));
      // opacityValue = 1 - (distance/10000);
      this.#ctx.strokeStyle = this.color;
      this.#ctx.beginPath();
      this.#ctx.lineWidth = 1.5;
      this.#ctx.moveTo(this.particleArray[i].x, this.particleArray[i].y);
      this.#ctx.lineTo(this.particleArray[i+1].x, this.particleArray[i+1].y);
      this.#ctx.stroke();
    }
  }
}
